apps/vmq_server/src/vmq_reg_redis_trie.erl

```diff
@@ -402,9 +412,196 @@ trie_delete_path(MP, [{Node, Word, _} | RestPath]) ->
     Edge = #trie_edge{node_id = NodeId, word = Word},
     ets:delete(vmq_redis_trie, Edge),
     case ets:lookup(vmq_redis_trie_node, NodeId) of
-        [_] ->
-            ets:delete(vmq_redis_trie_node, NodeId),
-            trie_delete_path(MP, RestPath);
+        [TrieNode = #trie_node{edge_count = EdgeCount}] ->
+            case EdgeCount > 1 of
+                true ->
+                    ets:insert(
+                        vmq_redis_trie_node,
+                        TrieNode#trie_node{edge_count = EdgeCount - 1}
+                    );
+                false ->
+                    ets:delete(vmq_redis_trie_node, NodeId),
+                    trie_delete_path(MP, RestPath)
+            end;
         [] ->
             ignore
     end.
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+%%%%%%%%%%%%%
+%%% Tests
+%%%%%%%%%%%%%
+
+complex_acl_test_() ->
+    [
+        {"Complex ACL Test - Add complex topic",
+            {setup, fun setup_vmq_reg_redis_trie/0, fun teardown/1, fun add_complex_acl_test/1}},
+        {"Complex ACL Test - Delete complex topic",
+            {setup, fun setup_vmq_reg_redis_trie/0, fun teardown/1, fun delete_complex_acl_test/1}},
+        {"Complex ACL Test - Sub-topic whitelisting",
+            {setup, fun setup_vmq_reg_redis_trie/0, fun teardown/1, fun subtopic_subscribe_test/1}},
+        {"Complex ACL Test - Get individual complex topics",
+            {setup, fun setup_vmq_reg_redis_trie/0, fun teardown/1, fun get_complex_topics_test/1}}
+    ].
+
+setup_vmq_reg_redis_trie() ->
+    TABLE_OPTS = [public, named_table, {read_concurrency, true}],
+    ok = application:set_env(vmq_enhanced_auth, enable_acl_hooks, true),
+    ok = application:set_env(vmq_server, default_reg_view, vmq_reg_redis_trie),
+    vmq_enhanced_auth:init(),
+    ets:new(topic_labels, [named_table, public, {write_concurrency, true}]),
+    ets:new(vmq_redis_trie_node, [{keypos, 2} | TABLE_OPTS]),
+    ets:new(vmq_redis_trie, [{keypos, 2} | TABLE_OPTS]),
+    vmq_reg_redis_trie.
+
+teardown(RegView) ->
+    case RegView of
+        vmq_reg_redis_trie ->
+            ets:delete(vmq_redis_trie),
+            ets:delete(vmq_redis_trie_node);
+        _ ->
+            ok
+    end,
+    ets:delete(vmq_enhanced_auth_acl_read_all),
+    ets:delete(vmq_enhanced_auth_acl_write_all),
+    ets:delete(vmq_enhanced_auth_acl_read_user),
+    ets:delete(vmq_enhanced_auth_acl_write_user),
+    ets:delete(topic_labels),
+    application:unset_env(vmq_enhanced_auth, enable_acl_hooks),
+    application:unset_env(vmq_server, default_reg_view).
+
+add_complex_acl_test(_) ->
```

Comment:
Please add unit tests. This is integration test. And since this is testing ACLs for complex topics, we can move this to enhanced_auth plugin only.

Reply:
Added the unit test for individual functions. and moved the integration test back to enhanced_auth_plugin.